firedrake/slate/slac/utils.py

```diff
@@ -136,6 +151,213 @@ def visit_Symbol(self, o, *args, **kwargs):
+                first_ind += (index,)
+            else:
+                first_ind += (index.start,)
+        first_block = tuple(range(ten, ten+1) for ten in first_ind)
+        index_offset = ()
+
+        # i points to the dim
+        # idx points to the shape of that block matrix in that dim
+        for i, idx in enumerate(first_block):
+            if idx.start == 0:
+                index_offset += (0, )
+            else:
+                index_offset += ((sum(A.shapes[i][:idx.start])), )
+
+        index_extent = tuple(sum(shape) for shape in tensor.shapes.values())
+        self.builder.create_index(index_extent, tensor)
+        gem_index = self.builder.gem_indices[tensor]
+
+        dim2idxs = []
+        for i, dim in enumerate(tensor.shapes.keys()):
+            dim2idxs += ((index_offset[i], ((gem_index[i], 1), )),)
+
+        return ComponentTensor(FlexiblyIndexed(node, tuple(dim2idxs)), gem_index)
+
+    def index_tensor(self, var, idx):
+        """ This method is a helper to:
+        - pick up free indices from a variable in the right
+        order
+        - if needed (it is for IndexSum and Indexed),
+        pull up a scalar variable to a tensor thing with this indices
+        - index it with new indices :arg idx.
+        """
+        if type(var) == Variable or type(var) == ComponentTensor or type(var) == Inverse or type(var) == Solve:
+            var = Indexed(var, idx)
+        else:
+            assert True, "Variable type is "+str(type(var))+". Must be a type that has shape."
+        return var
+
+    @slate_to_gem.register(sl.Solve)
+    def slate_to_gem_solve(self, tensor, node_dict):
+        fac, B = tensor.operands  # TODO is first operand always factorization?
+        A, = fac.operands
+        self.builder.create_index(A.shape, str(tensor)+"readssolve")
+        A_indices = self.builder.gem_indices[str(tensor)+"readssolve"]
+        self.builder.create_index(B.shape, str(tensor)+"readsbsolve")
+        B_indices = self.builder.gem_indices[str(tensor)+"readsbsolve"]
+        ret_A = ComponentTensor(self.index_tensor(node_dict[A], A_indices), A_indices)
+        ret_B = ComponentTensor(self.index_tensor(node_dict[B], B_indices), B_indices)
+        ret = Solve(ret_A, ret_B)
+        return ret
+
+    @slate_to_gem.register(sl.Inverse)
+    def slate_to_gem_inverse(self, tensor, node_dict):
+        A, = tensor.operands
+        self.builder.create_index(A.shape, str(A)+"readsinv")
+        A_indices = self.builder.gem_indices[str(A)+"readsinv"]
+        ret = ComponentTensor(self.index_tensor(node_dict[A], A_indices), A_indices)
+        ret = Inverse(ret)
+        return ret
+
+    # TODO how do we deal with surpressed factorization nodes,
+    # maybe populate decompdict and pass through to loopy later?
+    @slate_to_gem.register(sl.Factorization)
+    def slate_to_gem_factorization(self, tensor, node_dict):
+        self.decomp_dict.setdefault(tensor, tensor.decomposition)
+        return []
```

Comment:
With the move to not naming indices here (and the changes to the Inverse and Solve nodes) I think this can be much shorter. I think everything has a straightforward translation, with `Tensor` and `AssembledVector` being translated to `gem.Variable` of appropriate shape.

So I would do something like:

```python
from gem.node import Memoizer
import gem

@singledispatch
def _slate2gem(expr, self):
    raise AssertionError

@_slate2gem.register(slate.Tensor)
@_slate2gem.register(slate.AssembledVector)
def _slate2gem_tensor(expr, self):
    shape = expr.shape
    return self.var2terminal.setdefault(gem.Variable(name, shape), expr)

@_slate2gem.register(slate.Inverse)
def _slate2gem_inverse(expr, self):
    return gem.Inverse(*map(self, expr.children))

...

def slate2gem(expressions):
    mapper = Memoizer(_slate2gem)
    mapper.var2terminal = OrderedDict()
    return list(map(mapper, expressions))
```

Now you have the gem expression for the tensor algebra and you've got the information about which variables correspond to which UFL forms (terminal tensors).